relay_test.go

```diff
@@ -1001,3 +1001,132 @@ func TestRelayRaceCompletionAndTimeout(t *testing.T) {
 		wg.Wait()
 	})
 }
+
+func TestRelayArg2OffsetIntegration(t *testing.T) {
+	const (
+		testRK        = "routingkey"
+		testMethod    = "echo"
+		wantArg2Start = len(testRK) + len(testMethod) + 70 /*data before arg1*/
+		payloadLeft   = MaxFramePayloadSize - wantArg2Start
+
+		arg2Data = "moe"
+	)
+
+	ctx, cancel := NewContext(testutils.Timeout(time.Second))
+	defer cancel()
+
+	inspector := newFailedRelayHost()
+	opts := testutils.NewOpts().
+		SetRelayOnly().
+		SetRelayHost(inspector)
+
+	ch := testutils.NewServer(t, opts)
+	defer ch.Close()
+
+	client := testutils.NewClient(t, nil /*opts*/)
+	defer client.Close()
+
+	tests := []struct {
+		msg               string
+		arg2Data          string
+		arg2Flush         bool
+		arg2PostFlushData string
+		wantEndOffset     int
+		wantHasMore       bool
+	}{
+		{
+			msg:           "all within a frame",
+			arg2Data:      arg2Data,
+			wantEndOffset: wantArg2Start + len(arg2Data),
+		},
+		{
+			msg:           "arg2 flushed",
+			arg2Data:      arg2Data,
+			arg2Flush:     true,
+			wantEndOffset: wantArg2Start + len(arg2Data),
+			wantHasMore:   true,
+		},
+		{
+			msg:               "arg2 flushed called then write again",
+			arg2Data:          arg2Data,
+			arg2Flush:         true,
+			arg2PostFlushData: "more data",
+			wantEndOffset:     wantArg2Start + len(arg2Data),
+			wantHasMore:       true,
+		},
+		{
+			msg:           "no arg2",
+			wantEndOffset: wantArg2Start,
+		},
+		{
+			msg:           "XL arg2 which is fragmented",
+			arg2Data:      string(make([]byte, MaxFrameSize+100)),
+			wantEndOffset: wantArg2Start + payloadLeft,
+			wantHasMore:   true,
+		},
+		{
+			msg:           "large arg2 with 3 bytes left for arg3",
+			arg2Data:      string(make([]byte, payloadLeft-3)),
+			wantEndOffset: wantArg2Start + payloadLeft - 3,
+			wantHasMore:   false,
+		},
+		{
+			msg:           "large arg2, 2 bytes left",
+			arg2Data:      string(make([]byte, payloadLeft-2)),
+			wantEndOffset: wantArg2Start + payloadLeft - 2,
+			wantHasMore:   true, // arg3 is fragmented
```

Comment:
when there's 2 bytes left in the payload, do we not send arg3 length of 0? i think regardless of the length of arg3, we'd always set the "hasMore" flag but only have arg3 on the next frame

Reply:
Good point - I think there is a 2 byte mistake, as I forgot to account for the 2 bytes Arg2 length field; writing payloadLeft-2 for arg2 actually means we will write

2 (arg2Length) + (payloadLeft - 2) = payloadLeft

i.e. we used up all payload space for arg2. Hence the frame is fragmented.

Reply:
Actually, we don't use the 2 bytes in current implementation. Validated it by using tcpdump and wireshark.